src/prefect_server/api/states.py

```diff
@@ -100,6 +123,101 @@ async def set_flow_run_state(
             return_exceptions=True,
         )
 
+    # --------------------------------------------------------
+    # Queueing runs if using concurency limits
+    # --------------------------------------------------------
+
+    # If the run is not labeled or we've already checked during,
+    # a previous recursive call, we avoid this code branch
+    # as it's manually injecting a delay (and it isn't concurrency limited)
+    if (
+        (state.is_running() or state.is_submitted())
+        and flow_run.labels
+        and not has_been_concurrency_checked
+    ):
+        # Flow Concurrency Limits
+        # If the run is already occupying a slot (Submitted -> Running)
+        # or is attempting to occupy a slot (X -> Submitted)
+        existing_state = state_schema.load(flow_run.serialized_state)
+
+        # Due to the Agent process submitting multiple runs from Scheduled
+        # to Submitted at once, we need to acquire a lock to both check &
+        # set the new state if there are concurrency limits in place to
+        # avoid race conditions between the time a check has occurred and
+        # when an update has occurred.
+        server_context = context.get_context()
+        flow_concurrency_lock = server_context.get(
```

Comment:
An alternate implementation doesn't need to rely on the `flow_concurrency_lock` or `has_been_concurrency_checked` parameters. See comment in GraphQL code about `flow_concurrency_lock` reason.

`has_been_concurrency_checked` was initially added to avoid needing to manually manage the lock within the method (with the assumption that a differing function signature was not a problem). We can revert the signature back to the cloud compatible one, then manage acquiring the lock before the concurrency check and release it after the state insert. Coincidentally, this would also be better for performance since we'd avoid querying the run info again.